### PR TITLE
Add orchestration and entity context type declarations

### PR DESCRIPTION
**Pull request by hossam-nasr**
Part of #416:

* Adds a `DurableOrchestrationContext` class declaration exported from the root of the package that the `DurableOrchestrationContext` class implements. This is the type of the `context.df` object in orchestrations.
* Converts `DurableEntityContext` type into an interface, and also exports it from the root of the package. This is the type of the `context.df` object in entity functions

### CODE REVIEW

File: src/durableorchestrationcontext.ts
Comment (davidmrdavid) [inline src/durableorchestrationcontext.ts]: Seems we removed the DurableEntityContext file. Why?

**Review by davidmrdavid — COMMENTED** [review]
PR makes sense to me except for the deletion of the entity context:

File: src/durableorchestrationcontext.ts
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts]: It was all just moved to the `entityTypes.ts` file, to get us one step closer to #458. The plan is to have all the types exported from the `types` directory, then convert the `.ts` files there into `.d.ts` files, and that will now become the only source `.d.ts` files.

**Review by davidmrdavid — APPROVED** [review]
🚢 it

File: src/testingUtils.ts
Comment (davidmrdavid) [inline src/testingUtils.ts:11]: I just realized that it's a bit confusing to have a `DurableEntityContext` and an `EntityContext` type. What is the difference between the two again?

File: src/durableorchestrationcontext.ts
Comment (davidmrdavid) [inline src/durableorchestrationcontext.ts:41]: so there's two `DurableOrchestrationContext` types now? I'm a bit confused by that.

File: src/types/entityTypes.ts
Comment (davidmrdavid) [inline src/types/entityTypes.ts:46]: I noticed this used to be an `export type` but now is `export interface`. Can you please clarify the difference?
Also, why does this export an `interface` where the orchestrator-context exports a `class`?

**Review by davidmrdavid — CHANGES_REQUESTED** [review]
Actually, on closer inspection, I have a few more questions :)

File: src/testingUtils.ts
Comment (hossam-nasr) [reply src/testingUtils.ts:11]: The `EntityContext` type is the type of the `context` argument passed to entity functions, and `DurableEntityContext` is the type of the `df` object on the `EntityContext`. If you have better suggestions for the naming, I'm open to it, but changing `DurableEntityContext` would be a breaking change. Ofc, this can be avoided if we implement #422 😛

File: src/durableorchestrationcontext.ts
Comment (hossam-nasr) [reply src/durableorchestrationcontext.ts:41]: No, this file hosts the implementation of the `DurableOrchestrationContext` class, which `implements` the `DurableOrchestrationContext` _declaration_ in the `types` directory. It's declaration/interface vs implementation. The plan is that the class declaration defined in the `types` directory will become our public "contract" for the API provided by the `DurableOrchestrationContext`, which _excludes_ things like private fields and methods, otherwise included in the automatically generated `.d.ts` files

File: src/types/entityTypes.ts
Comment (hossam-nasr) [reply src/types/entityTypes.ts:46]: For this use-case, there are no relevant differences. I'm not sure what the subtle differences between `type` and `interface` in TypeScript (at least `>v2.1`) are, but for defining object/class structure, they're effectively the same. However, it feels to me at least that using `interface` for this use-case is more idiomatic, when defining the structure of an object. I think the same also applies to the earlier `DurableClient` and `DurableOrchestrationContext` classes, and if it was up to me, I'd rather change those to interfaces too. One argument is that you, as the user, _shouldn't_ be calling the constructor of those types. However, changing classes to interfaces _would_ be a breaking change, which is why I haven't done that here, but changing a `type` to an `interface` isn't.

File: src/testingUtils.ts
Comment (davidmrdavid) [reply src/testingUtils.ts:11]: I see. Yeah the naming here is not ideal but I'm not ready to change this at this time :) . Let's put a pin on this, but thanks for clarifying!

**Review by davidmrdavid — APPROVED** [review]
LGTM after discussion

File: src/types/entityTypes.ts
Comment (ejizba) [reply src/types/entityTypes.ts:46]: I agree with David's sentiment here and would prefer if we avoid changes like this during two scenarios:
1. When dealing with our public contract. If the risk and the value are equal (both low in this case), I err on the side of caution
2. During PRs that are mostly just moving code. These PRs are difficult to review, so even minor changes should be avoided

That being said, I don't have a specific example of when switching a type to an interface would cause a breaking change so I won't block on this PR. Please consider this feedback in future PRs, though

File: src/types/orchestrationTypes.ts
Comment (ejizba) [inline src/types/orchestrationTypes.ts:94]: Same feedback as https://github.com/Azure/azure-functions-durable-js/pull/450#discussion_r1100574673

This comment is the only reason I'm requesting changes

File: src/types/orchestrationTypes.ts
Comment (ejizba) [inline src/types/orchestrationTypes.ts:267]: I appreciate that you cleaned up the docs in several places, but in the future please avoid _any_ change that you possibly can when moving code. This is to speed up the review of these PRs and ensure things don't slip in on accident

**Review by ejizba — CHANGES_REQUESTED** [review]

File: src/types/entityTypes.ts
Comment (hossam-nasr) [reply src/types/entityTypes.ts:46]: Sounds good, I appreciate the feedback, thank you!

File: src/types/orchestrationTypes.ts
Comment (hossam-nasr) [reply src/types/orchestrationTypes.ts:94]: This is the type signature of the constructor:

```TS
constructor(
        state: HistoryEvent[],
        instanceId: string,
        currentUtcDateTime: Date,
        isReplaying: boolean,
        parentInstanceId: string | undefined,
        longRunningTimerIntervalDuration: string | undefined,
        maximumShortTimerDuration: string | undefined,
        defaultHttpAsyncRequestSleepTimeMillseconds: number | undefined,
        schemaVersion: ReplaySchema,
        input: unknown,
        private taskOrchestratorExecutor: TaskOrchestrationExecutor
    );
```

Exposing the `constructor()` would also force us to expose other types like `HistoryEvent` (which in turn exposes `HistoryEventType`), `ReplaySchema`, `TaskOrchestrationExecutor` (which in turn also exposes `TaskBase` and other internal classes), etc. In the interest of #352, and in the spirit of this change to let us more intentionally the public APIs, I think we should omit the constructor. Users shouldn't be using this constructor anyway.

File: src/types/orchestrationTypes.ts
Comment (hossam-nasr) [reply src/types/orchestrationTypes.ts:267]: Sounds good, thank you for the feedback.

**Review by ejizba — APPROVED** [review]

File: src/types/orchestrationTypes.ts
Comment (hossam-nasr) [reply src/types/orchestrationTypes.ts:94]: Added constructor, will file issue to remove

**Pull request merged**
